libpod/options.go

```diff
@@ -1057,6 +1061,44 @@ func WithLogPath(path string) CtrCreateOption {
 	}
 }
 
+// WithLogPolicy sets the logging policy.
+func WithLogPolicy(policy string) CtrCreateOption {
+	return func(ctr *Container) error {
+		if ctr.valid {
+			return define.ErrCtrFinalized
+		}
+		switch policy {
+		case "":
+			return errors.Wrapf(define.ErrInvalidArg, "log policy must be set")
+		case "backpressure", "drop", "ignore", "passthrough":
+			break
+		default:
+			return errors.Wrapf(define.ErrInvalidArg, "invalid log policy")
+		}
+		ctr.config.LogPolicy = policy
+		return nil
+	}
+}
+
+// WithLogRateLimit sets the logging rate limit.
+func WithLogRateLimit(rateLimit string) CtrCreateOption {
+	return func(ctr *Container) error {
+		if ctr.valid {
+			return define.ErrCtrFinalized
+		}
+		if rateLimit == "" {
+			return errors.Wrapf(define.ErrInvalidArg, "log rate limit must be set")
```

Comment:
Do we need the rate limit specified if the policy is `ignore` or `passthrough`?

Reply:
A rate limit sure makes no sense for ignore or passthrough (which is the current unrestricted behavior, btw). How do you think should the CLI behave here, error out in that case, ignore, warn?

Reply:
So with no changes by the user, they will get `passthrough`.  So that case should work today, which I think this code as it stands will raise an error.

If they specify `ignore` or `passthrough`, then they should not specify a rate, and it should be an error if they do.

If they specify `backpressure` or `drop`, then they must specify the rate.

Reply:
Alright, makes perfect sense.

Reply:
Added the checks to https://github.com/syedriko/libpod/blob/004258dd4343d055876137d490ca3b2cbff68f15/pkg/spec/createconfig.go#L290

Reply:
Backpressure/drop should be allowed without a rate - that's the "write as fast as the pipe will let you " case, which IMO is one of the most useful behaviours - responds to the actual runtime rate the system can take logs rather than a rate somebody calculated/guessed in advance.

Reply:
I agree, but at the moment, we don't have any downstream mechanism with natural backpressure, we went with having to specify the rate for now and reevaluate once we have such mechanism.

Reply:
We're writing to files and the file writes block until the file is written, so  the current default is "backpressure" with no rate-limit. I think you could simplify validation so that rate-limit is optional regardless of policy. The "ignore" policy trivially respects any rate-limit (even 0) There is no such thing as "passthrough" - the existing behaviour is backpressure with no rate-limit.

Reply:
The semantics starts to get interesting if we consider where the backpressure comes from, from the code we're looking at, where the rate would sure apply, or downstream from this code. In that frame of reference, how would we implement drop without a rate limit, by making the log fd non-blocking, etc?

Reply:
A non-blocking partial write followed by a rate calculation I'd say. 

The whole thing would be simpler as a poll loop with non-blocking reads/writes and a timer fd.

Reply:
Without a rate limit there's no basis for calculation though. A simple implementation can attempt to write to the non-blocking log fd and drop on seeing EAGAIN. A more involved one could poll() the fd and flip a "drop" flag.

Reply:
For backpressure there's nothing to calculate, you just block on write and delay reading as a consequence - the downstream determines how fast you go. For drop what you describe is right, but I would add a small buffer to avoid dropping due to minor jitters in read/write rate. Then drop on read with buffer-full.